src/extractor/way_restriction_map.cpp

```diff
@@ -0,0 +1,199 @@
+#include "extractor/way_restriction_map.hpp"
+
+#include <iterator>
+#include <tuple>
+#include <utility>
+
+namespace osrm
+{
+namespace extractor
+{
+
+namespace
+{
+struct FindViaWay
+{
+    bool operator()(const std::tuple<NodeID, NodeID> value,
+                    const TurnRestriction &restriction) const
+    {
+        const auto &way = restriction.AsWayRestriction();
+        return value < std::tie(way.in_restriction.via, way.out_restriction.via);
+    }
+    bool operator()(const TurnRestriction &restriction,
+                    const std::tuple<NodeID, NodeID> value) const
+    {
+        const auto &way = restriction.AsWayRestriction();
+        return std::tie(way.in_restriction.via, way.out_restriction.via) < value;
+    }
+};
+
+} // namespace
+
+WayRestrictionMap::WayRestrictionMap(const std::vector<TurnRestriction> &turn_restrictions)
+{
+    // get all way restrictions
+    const auto extract_restrictions = [this](const auto &turn_restriction) {
+        if (turn_restriction.Type() == RestrictionType::WAY_RESTRICTION)
+        {
+            const auto &way = turn_restriction.AsWayRestriction();
+
+            // so far we can only handle restrictions that are not interrupted
+            if (way.in_restriction.via == way.out_restriction.from &&
+                way.in_restriction.to == way.out_restriction.via)
+                restriction_data.push_back(turn_restriction);
+        }
+    };
+    std::for_each(turn_restrictions.begin(), turn_restrictions.end(), extract_restrictions);
+
+    const auto as_duplicated_node =
+        [](auto const &restriction) -> std::tuple<NodeID, NodeID, NodeID> {
+        auto &way = restriction.AsWayRestriction();
+        // group restrictions by the via-way. On same via-ways group by from
+        return std::make_tuple(
+            way.in_restriction.via, way.out_restriction.via, way.in_restriction.from);
+    };
+
+    const auto by_duplicated_node = [&](auto const &lhs, auto const &rhs) {
+        return as_duplicated_node(lhs) < as_duplicated_node(rhs);
+    };
+
+    std::sort(restriction_data.begin(), restriction_data.end(), by_duplicated_node);
+
+    std::size_t index = 0, duplication_id = 0;
+    // map all way restrictions into access containers
+    const auto prepare_way_restriction = [this, &index, &duplication_id, as_duplicated_node](
+        const auto &restriction) {
+        const auto &way = restriction.AsWayRestriction();
+        restriction_starts.insert(
+            std::make_pair(std::make_pair(way.in_restriction.from, way.in_restriction.via), index));
+        ++index;
+    };
+    std::for_each(restriction_data.begin(), restriction_data.end(), prepare_way_restriction);
```

Comment:
This should be a for loop for sure as it keep external state. Also `as_duplicate_node` is bound but not used, and I don't see a reference to `duplication_id`.